consensus/src/pending_order_votes.rs

```diff
@@ -0,0 +1,176 @@
+
+                return OrderVoteReceptionResult::EquivocateVote;
+            }
+        }
+
+        self.author_to_order_vote.insert(
+            (order_vote.author(), round),
+            (order_vote.clone(), li_digest),
+        );
+
+        let len = self.li_digest_to_votes.len() + 1;
+        // obtain the ledger info with signatures associated to the order vote's ledger info
+        let (_hash_index, status, li_with_sig) =
+            self.li_digest_to_votes.entry(li_digest).or_insert_with(|| {
+                // if the ledger info with signatures doesn't exist yet, create it
+                (
+                    len,
+                    OrderVoteStatus::NotEnoughVotes,
+                    LedgerInfoWithPartialSignatures::new(
+                        order_vote.ledger_info().clone(),
+                        PartialSignatures::empty(),
+                    ),
+                )
+            });
+        let validator_voting_power = validator_verifier
+            .get_voting_power(&order_vote.author())
+            .unwrap_or(0);
+        if validator_voting_power == 0 {
+            warn!(
+                "Received vote with no voting power, from {}",
+                order_vote.author()
+            );
+        }
+        li_with_sig.add_signature(order_vote.author(), order_vote.signature().clone());
+
+        // check if we have enough signatures to create a QC
+        match validator_verifier.check_voting_power(li_with_sig.signatures().keys(), true) {
+            // a quorum of signature was reached, a new QC is formed
+            Ok(aggregated_voting_power) => {
+                assert!(
+                    aggregated_voting_power >= validator_verifier.quorum_voting_power(),
+                    "QC aggregation should not be triggered if we don't have enough votes to form a QC"
+                );
+                match li_with_sig.aggregate_signatures(validator_verifier) {
+                    Ok(ledger_info_with_sig) => {
+                        *status = OrderVoteStatus::EnoughVotes;
+                        OrderVoteReceptionResult::NewLedgerInfoWithSignatures(Arc::new(
+                            ledger_info_with_sig,
+                        ))
+                    },
+                    Err(e) => OrderVoteReceptionResult::ErrorAggregatingSignature(e),
+                }
+            },
+
+            // not enough votes
+            Err(VerifyError::TooLittleVotingPower { voting_power, .. }) => {
+                OrderVoteReceptionResult::VoteAdded(voting_power)
+            },
+
+            // error
+            Err(error) => {
+                error!(
+                    "MUST_FIX: order vote received could not be added: {}, order vote: {}",
+                    error, order_vote
+                );
+                OrderVoteReceptionResult::ErrorAddingVote(error)
+            },
+        }
+    }
+
+    // Removes votes older than round-1
+    pub fn set_round(&mut self, round: u64) {
```

Comment:
I think we should garbage collect order votes based on the ordered round (`highest_committed_round`), not when entering the new round.

Reply:
it's probably better to make this live outside of round state but in round manager so it can track multiple rounds and garbage collect itself too

Reply:
I changed the code to do garbage collection based on `highest_committed_round`. I kept pending order votes still in round state as didn't make much difference.